### PR TITLE
test: Add ghost variables and inline assertions (SC-485)

### PR DESCRIPTION
**Pull request by lucas-manuel**

### CODE REVIEW

File: test/invariant/Invariants.t.sol [hunk 1]
```diff
@@ -60,14 +61,104 @@ abstract contract PSMInvariantTestBase is PSMTestBase {
...
+        }
+
+        // Loop twice to avoid underflows between LPs
+        for(uint256 i; i < 3; i++) {
+            address swapper = swapperHandler.swappers(i);
+
+            totalValueSwappedOut +=
+                swapperHandler.swapsIn(swapper, address(usdc)) * 1e12 +
+                swapperHandler.swapsIn(swapper, address(dai)) +
+                swapperHandler.swapsIn(swapper, address(sDai)) * rateProvider.getConversionRate() / 1e27;
```
Comment (hexonaut) [inline test/invariant/Invariants.t.sol:158]: These should all be `swapsOut`.

File: test/invariant/Invariants.t.sol [hunk 2]
```diff
@@ -352,14 +511,17 @@ contract PSMInvariants_TimeBasedRateSetting_NoTransfer is PSMInvariantTestBase {
...
         targetContract(address(lpHandler));
         targetContract(address(swapperHandler));
         targetContract(address(timeBasedRateHandler));
+
+        // Check that LPs used for swap assertions are correct to not get zero values
+        assertEq(swapperHandler.lp0(), lpHandler.lps(0));
     }
 
-    function invariant_A() public view {
+    function invariant_A_test() public view {
```
Comment (hexonaut) [inline test/invariant/Invariants.t.sol:524]: Should be `invariant_A`.

File: test/invariant/handlers/RateSetterHandler.sol [hunk 3]
```diff
@@ -1,29 +1,50 @@
...
+        // 2. Cache starting state
+        uint256 startingConversion = psm.convertToShares(1e18);
+        uint256 startingValue      = psm.totalAssets();
+
+        // 3. Perform action against protocol
         rateProvider.__setConversionRate(rate);
 
+        // 4. Perform action-specific assertions
+        assertGe(
+            psm.convertToAssetValue(1e18) + 1,
```
Comment (hexonaut) [inline test/invariant/handlers/RateSetterHandler.sol:36]: Should this not be a comparison with `convertToShares` to match the units of `startingConversion`?

File: test/invariant/handlers/TimeBasedRateHandler.sol [hunk 4]
```diff
@@ -1,46 +1,96 @@
...
+        // 3. Perform action against protocol
         dsrOracle.setPotData(IDSROracle.PotData({
             dsr: uint96(dsr),
             chi: uint120(chi),
             rho: uint40(rho)
         }));
+
+        // 4. Perform action-specific assertions
+        assertGe(
+            psm.convertToAssetValue(1e18) + 1,
```
Comment (hexonaut) [inline test/invariant/handlers/TimeBasedRateHandler.sol:53]: Again, should this be `convertToShares` to match above?

File: test/invariant/handlers/TimeBasedRateHandler.sol [hunk 5]
```diff
@@ -1,46 +1,96 @@
...
+        // 2. Cache starting state
+        uint256 startingConversion = psm.convertToShares(1e18);
+        uint256 startingValue      = psm.totalAssets();
+
+        // 3. Perform action against protocol
+        skip(warpTime);
+
+        // 4. Perform action-specific assertions
+        assertGe(
+            psm.convertToAssetValue(1e18),
```
Comment (hexonaut) [inline test/invariant/handlers/TimeBasedRateHandler.sol:81]: Same here.

File: test/invariant/handlers/TransferHandler.sol [hunk 6]
```diff
@@ -9,29 +9,64 @@ import { PSM3 } from "src/PSM3.sol";
...
-
         vm.prank(sender);
         asset.transfer(address(psm), amount);
 
+        // 4. Update ghost variable(s)
+        transfersIn[address(asset)] += amount;
+
+        // 5. Perform action-specific assertions
+        assertGe(
+            psm.convertToAssetValue(1e18) + 1,
```
Comment (hexonaut) [inline test/invariant/handlers/TransferHandler.sol:58]: Same here.

**Review by hexonaut — CHANGES_REQUESTED** [review]

File: test/invariant/Invariants.t.sol (hunk 2 above)
Comment (lucas-manuel) [reply test/invariant/Invariants.t.sol:524]: Updated

File: test/invariant/Invariants.t.sol (hunk 1 above)
Comment (lucas-manuel) [reply test/invariant/Invariants.t.sol:158]: Ah yep updated

Comment (lucas-manuel) [reply test/invariant/Invariants.t.sol:158]: Causing issues, debugging

File: test/invariant/handlers/RateSetterHandler.sol (hunk 3 above)
Comment (lucas-manuel) [reply test/invariant/handlers/RateSetterHandler.sol:36]: Updated to use `convertToAssetValue` for both

File: test/invariant/handlers/TimeBasedRateHandler.sol (hunk 4 above)
Comment (lucas-manuel) [reply test/invariant/handlers/TimeBasedRateHandler.sol:53]: Updated to use `convertToAssetValue` for both

File: test/invariant/handlers/TimeBasedRateHandler.sol (hunk 5 above)
Comment (lucas-manuel) [reply test/invariant/handlers/TimeBasedRateHandler.sol:81]: Updated to use `convertToAssetValue` for both

File: test/invariant/handlers/TransferHandler.sol (hunk 6 above)
Comment (lucas-manuel) [reply test/invariant/handlers/TransferHandler.sol:58]: Updated to use `convertToAssetValue` for both

File: test/invariant/Invariants.t.sol [hunk 7]
```diff
@@ -477,7 +479,7 @@ contract PSMInvariants_RateSetting_WithTransfers is PSMInvariantTestBase {
         _checkInvariant_E();
     }
 
-    function invariant_F() public view {
+    function skip_invariant_F() public view {
```
Comment (hexonaut) [inline test/invariant/Invariants.t.sol:482]: This is a new pattern or a mistake?

**Review by hexonaut — CHANGES_REQUESTED** [review]
Code looks fine now. I've got passing on my local machine, but CI is failing.

**Review by hexonaut — DISMISSED** [review]
LGTM

File: test/invariant/handlers/TimeBasedRateHandler.sol [hunk 8]
```diff
@@ -12,7 +12,7 @@ contract TimeBasedRateHandler is HandlerBase, StdCheats {
 
     uint256 public dsr;
 
-    uint256 constant ONE_HUNDRED_PCT_APY_DSR = 1.000000021979553151239153027e27;
+    uint256 constant TWENTY_HUNDRED_PCT_APY_DSR = 1.000000005781378656804591712e27;
```
Comment (hexonaut) [inline test/invariant/handlers/TimeBasedRateHandler.sol:15]: `TWENTY_PCT_APY_DSR`

**Review by hexonaut — APPROVED** [review]

**Pull request merged**